Properly support interactive Subversion client

johnthagen opened the pull request:
Fixes #6386

cc @cjerdonek

johnthagen commented:
@cjerdonek Do you know of any public Internet-facing SVN repos that contain Python packages we could use to test with? I have some private ones I can test against, but it would also be helpful if there were some that others (like you) could test as part of the review. Even if they aren't private (e.g. needing a password), I think it would be a helpful smoke test.

cjerdonek commented:
> Do you know of any public Internet-facing SVN repos that contain Python packages we could use to test with?

I don't know offhand, but I agree it would be good to find some examples.

On your PR, for caching reasons, the methods needing `get_remote_call_options()` are going to have to become instance methods again. Can you tell if that will be a manageable change?

cjerdonek commented on tests/functional/test_install_vcs_svn.py:
```diff
@@ -4,9 +4,12 @@
-def test_obtain_should_recognize_auth_info_url(call_subprocess_mock, script):
+def test_obtain_should_recognize_auth_info_url(
+        call_subprocess_mock, get_remote_call_options_mock, script):
+    get_remote_call_options_mock.return_value = []
```
If you instantiate `Subversion` with `use_interactive=False`, you can feed two birds with one scone and both (1) avoid having to mock `get_remote_call_options()` and (2) test that remote call options are actually used.

johnthagen replied:
> feed two birds with one scone 

🤣

johnthagen replied:
Great idea, I've pushed in the update.

johnthagen commented:
@cjerdonek Some initial good news! I tested the following configurations with this PR:

- Ubuntu 18.04 LTS: `svn 1.9.7`
- Ubuntu 16.04 LTS: `svn 1.9.3`
- CentOS 7: `svn 1.7.14`

And `pip` successfully prompted the user for the password for the private Subversion repo in all situations! I saw `pip` log the use of `--force-interactive` for `svn 1.9.x` as expected and not for `svn 1.7.14` (again as expected).

So it would seem the logic is at least sound and actually solves the issue. (Very cool!)

cjerdonek commented:
Great to hear! Great work!

johnthagen commented:
> On your PR, for caching reasons, the methods needing get_remote_call_options() are going to have to become instance methods again.

Yeah, I noticed that too. It seems like this will bubble up into most of the `Subversion` methods, and also (more intrusively) into `VersionControl` base class (for example `get_requirement_revision`).

So, it looks pretty invasive (I'm also not very familiar with where this cacheing would happen. Does `pip` keep around a global `Subversion`/`Git`/etc object where this will be referenced?)

johnthagen commented:
The macOS failures look spurious and unrelated to this PR?

cjerdonek commented:
> (I'm also not very familiar with where this cacheing would happen. Does pip keep around a global Subversion/Git/etc object where this will be referenced?)

I was referring to the `self._vcs_version` caching that you did. But the caching you're asking about happens with `VcsSupport._registry`. That's a dictionary that stores an instance for each `VersionControl` base class once, globally. I made that change recently (to store instances rather than types) in the registry precisely to take advantage of the `self._vcs_version` caching you added. (Otherwise, `self._vcs_version` would have to be re-obtained each time the `Subversion` object is re-instantiated, which would defeat the purpose of the `_vcs_version` cache.)

For the class methods, if that's something you'd feel more comfortable me doing, let me know and I could do it. It should be an easy, more or less mechanical change. Also, depending on the size of the change, it should maybe go in a separate refactoring PR, too. A lot of those methods were made class methods before this use case came up.

cjerdonek commented:
> The macOS failures look spurious and unrelated to this PR?

Yes, those are flaky tests (I filed an issue about them not too long ago). And glad you like the "scone" idiom. :) I learned about that recently so want to use it more..

johnthagen commented:
> I made that change recently (to store instances rather than types) in the registry precisely to take advantage of the self._vcs_version caching you added.

Ah cool, that makes sense.

> For the class methods, if that's something you'd feel more comfortable me doing, let me know and I could do it.

Yeah, if you're willing I think it'd be more efficient for you to the the refactor since I'm less familiar with anything outside of the `Subversion` class.

> Also, depending on the size of the change, it should maybe go in a separate refactoring PR, too

Yeah, I feel it would be bets for that refactor to go into another PR. That way, this PR can be minimal for future posterity.

johnthagen commented:
I see you reported the flaky test in #6235 (referencing to cross-post another example of failure).

cjerdonek commented:
Almost all of the class method changes would be needed only for `_get_svn_url_rev()`, which is where `svn info` is called (and only for SVN >= 1.7 in that method). Are you sure that `svn info` needs to be called interactively, or can `_get_svn_url_rev()` be implemented in a way that doesn't require calling `svn info`? One reason I ask is that `_get_svn_url_rev()` is only needed to get (1) the current integer revision number, and (2) the remote url, which both seem like they should require access only to local information. (And for SVN < 1.7, `svn info` isn't needed to get the same information.)

cjerdonek commented on tests/functional/test_install_vcs_svn.py:
```diff
@@ -8,11 +8,11 @@
 def test_obtain_should_recognize_auth_info_url(call_subprocess_mock, script):
     url = 'svn+http://username:password@svn.example.com/'
-    svn = Subversion()
+    svn = Subversion(use_interactive=False)
```
I would add a comment (here and for the other one) saying you're passing `use_interactive=False` to check that the remote call options are being used.

johnthagen commented:
> are you sure that svn info needs to be called interactively

Oh, good catch. Looking at this more closely you're right, we don't need to call interactively since it's not running `svn info` *against* a URL (I mis-interpreted the meaning of `_get_svn_url_rev`).

Since we're running it against a local directory, the interactive password issue shouldn't be relevant to this use of `svn info`.

I'll update the PR accordingly.

cjerdonek commented:
Okay, cool. That means the class method changes will be minimal.

johnthagen commented:
I looked over the remaining usages of `get_remote_call_options()` and they all seem necessary (involve interaction with the server which may require authentication).

cjerdonek commented:
You might as well add a code comment above the svn info invocation saying why it’s not needed in that case (especially because it was initially unclear).

cjerdonek commented:
FYI, I posted PR #6519 for the `fetch_new()` change.

johnthagen commented:
> You might as well add a code comment above the svn info invocation saying why it’s not needed in that case (especially because it was initially unclear).

Good idea. Done.

cjerdonek commented on news/6386.bugfix:
```diff
@@ -0,0 +1 @@
+Support running Subversion client interactively on Subversion 1.8+.
```
I think you can be a little more complete in the description here and say something like, "Extend to Subversion 1.8+ the behavior of calling Subversion in interactive mode when pip is run interactively." This way it's clearer that (1) this behavior is present < 1.8, and (2) the user doesn't have to do anything to get this behavior.

cjerdonek commented on tests/functional/test_install_vcs_svn.py:
```diff
@@ -21,10 +23,12 @@ def test_obtain_should_recognize_auth_info_url(call_subprocess_mock, script):
-        svn.name, 'export', '--username', 'username', '--password',
-        'password', 'http://svn.example.com/',
+        svn.name, 'export', '--non-interactive', '--username', 'username',
+        '--password', 'password', 'http://svn.example.com/',
```
My only other comment on this PR is to add a test like this for the other two methods to ensure that the `get_remote_call_options()` are getting added for all four commands. Some suggestions for this--

1. It would be good to pull into a helper function the code common to the four cases. This way you can reduce boilerplate.
2. Since the SVN call is mocked, isn't it the case that these tests don't actually hit the network? If that's the case, the `@pytest.mark.network` decorators aren't needed.

johnthagen replied:
> Since the SVN call is mocked, isn't it the case that these tests don't actually hit the network?

Huh, yeah looks like this was erroneously added in c5024dc851fc8c5e688607f40c3a4a849617d563.

johnthagen replied:
> It would be good to pull into a helper function the code common to the four cases. This way you can reduce boilerplate.

I started to look at this. I agree there is some boilerplate, but it didn't seem too bad compared to normal testing boilerplate. Also since each test calls a different method, I couldn't come up with a way to reduce boilerplate meaningfully without making the tests harder to understand.

Could you review the latest PR and confirm you'd like this to be done (and perhaps how you think best to do it, if so)? Thanks.

cjerdonek replied:
The way I would do it is make a `TestSubversionArgs` `TestCase` class, and then use the approach at the end of this section to do the patching in `setUp()`: https://docs.python.org/3.5/library/unittest.mock-examples.html#applying-the-same-patch-to-every-test-method

Then I would add two helper methods to use at the beginning and end of each test:

1. `svn, dest, url = self.get_test_data()`
2. `assert_call_args(expected)`

johnthagen commented:
Build failure appears to be the same spurious macOS failure.

johnthagen commented:
@cjerdonek I believe I have addressed your comments.

Note that I switched away from using the `script` fixture as I couldn't find a way to get a normal pytest fixture to work with a `TestCase`. In this instance, since we're mocking `call_subprocess`, I don't think it matters, and actually makes the tests overall a little simpler.

cjerdonek commented on tests/functional/test_install_vcs_svn.py:
```diff
@@ -1,30 +1,69 @@
+    def test_fetch_new_should_include_remote_call_options(self):
+        self.svn.fetch_new(self.dest, self.url, self.rev_options)
+        self.assert_call_args(
+            [self.svn.name, 'checkout', '-q', '--non-interactive', self.url,
```
Generally, I think it's better to use strings you can see when asserting test expectations (e.g. use `"svn"` instead of `self.svn.name`). Otherwise, it's sort of like testing the test against itself, and there's no way to tell if you made a mistake in crafting the test input data.

johnthagen commented:
@cjerdonek I believe I have addressed all of your comments.

cjerdonek commented:
Looks great! Again, good job on the PR. 😀

cjerdonek commented:
Oops, I meant to click "squash" on that. Oh well, not a big deal.

johnthagen commented:
Thanks for all of your help getting this through @cjerdonek!

johnthagen commented:
>Oops, I meant to click "squash" on that. Oh well, not a big deal.

FYI, you probably already know this, but it is possible to set up the GitHub project to require squashed PRs, etc. if that would help.

pfmoore commented:
It's no big deal - pip doesn't require squashed PRs. I assume the only reason @cjerdonek intended to squash was because this PR has had quite a lot of revisions.

cjerdonek commented:
> I assume the only reason @cjerdonek intended to squash was because this PR has had quite a lot of revisions.

Yeah, basically.

> but it is possible to set up the GitHub project to require squashed PRs, etc. if that would help.

This occurred to me, too, but there are also many times we don't want to squash -- like when each commit is made separately meaningful and doesn't contain things getting corrected, etc.

As @pfmoore said, it's not a big deal. But if some kind of policy were possible, I think it'd be useful if the UI could be configured to prompt with an "are you sure?" message if you're not squashing and the PR has over N commits (with N perhaps 5 or so). But I doubt GitHub would have anything so fine-grained as that.

cjerdonek commented:
By the way, if you want to do a follow-up cleanup, you could move the `tests/functional/test_install_vcs_svn.py` tests into `unit/test_vcs.py` (since the former aren't actually functional but unit tests), and also move `export()`, `fetch_new()` to the end of the class (after the methods they depend on) since  they're currently mixed in with the class methods.

cjerdonek commented:
Hi @johnthagen, I just wanted to check again to see if you'd maybe be willing to do the two minor clean-ups I mentioned in my comment above. It would be appreciated!

johnthagen commented:
@cjerdonek Sure, I'll work on some PRs.
